src/NUnitFramework/framework/Constraints/ThrowsExceptionConstraint.cs

```diff
@@ -98,7 +98,12 @@ public override ConstraintResult ApplyTo<TActual>(TActual actual)
         public override ConstraintResult ApplyTo<TActual>(ActualValueDelegate<TActual> del)
         {
 #if ASYNC
-            if (typeof(TActual) == typeof(Task)) return ApplyTo(new AsyncTestDelegate(() => (Task)(object)del.Invoke()));
+            if (typeof(TActual) == typeof(Task))
```

Comment:
This (pre-existing) check is actually brittle. Folks are free to derive from Task and we would fail to recognize that.

I think this should be `if (typeof(Task).IsAssignableFrom(del.GetMethod("Invoke").ReturnType)) { ...`.
By doing this, the following code should work:

```c#
object obj = new Func<Task>(async () => await ...);
Assert.That(obj, Throws...);

// Task<T> derives from Task, so this is handled automatically.
obj = new Func<Task<T>>(async () => await ...);
Assert.That(obj, Throws...);
```

Reply:
Ah, missed that `del` is `ActualValueDelegate<TActual>` and not `TActual`. The above is code that would go in the `ApplyTo<TActual>(TActual actual)` overload.

For this overload, the check would simplify to `typeof(Task).IsAssignableFrom(typeof(Actual))` since `del.GetMethod("Invoke").ReturnType` will always be `typeof(Actual)` in this overload.

Reply:
Here is an example of why we should always use `typeof(Task).IsAssignableFrom`: `https://github.com/nunit/nunit/issues/3093#issuecomment-441184381`

@stevenaw If you do `if (typeof(Task).IsAssignableFrom(typeof(TActual)))`, a single `if` will take care of `Task`, Task<T>`, and `MyCustomTask : Task<T> { }`.

Reply:
Thanks @jnm2 . I think you might be looking at an old commit, I did that update here: https://github.com/nunit/nunit/pull/3087/commits/926a3b5e4423c76c0ee122cf035887af0c9984f1#diff-528b529b1298bbb4bd387f02090a3de3R101

:)

Reply:
Oh, perfect! Sorry about that! 😃